ts/ui/src/App.tsx

```diff
@@ -0,0 +1,162 @@
+import { Service } from "./service";
+import { AdlRequestError } from "./service/service-base";
+import { FetchHttp } from "./service/fetch-http";
+
+import * as API from "@protoapp/adl/protoapp/apis/ui";
+import React, { useState, useEffect, useCallback } from "react";
+
+const service = new Service(new FetchHttp(), "/api");
+
+// NOTE: This is a toy implementation of a client-side app that demonstrates the
+// use of a `Service` to make typesafe RPC requests from the ADL generated API-definition.
+
+const App: React.FC = () => {
+  // Login form state
+  const [email, setEmail] = useState<string>("");
+  const [password, setPassword] = useState<string>("");
+
+  // Auth state
+  const [accessToken, setAccessToken] = useState<string | null>(null);
+
+  // Messages state
+  const [messages, setMessages] = useState<API.Message[]>([]); // ADL defined API.Message
+  const [newMessage, setNewMessage] = useState<string>("");
+
+  const isLoggedIn = accessToken !== null;
+
+  const handleLogoutImmediate = useCallback(() => {
+    setAccessToken(null);
+    setMessages([]);
+  }, []);
+
+  const handleApiError = useCallback(
+    (err: unknown) => {
+      console.error(`API call failed`);
+      if (err instanceof AdlRequestError && err.respStatus === 401 && accessToken) {
+        console.log("Received 401, logging out.");
+        handleLogoutImmediate();
+      }
+    },
+    [accessToken, handleLogoutImmediate],
+  );
+
+  const fetchUserInfo = useCallback(async () => {
+    if (!isLoggedIn || !accessToken) return;
+    try {
+      await service.whoAmI(accessToken, null);
+    } catch (err) {
+      handleApiError(err);
+    }
+  }, [isLoggedIn, accessToken, handleApiError]);
```

Comment:
isLoggedIn usage in this function is redundant - presence of the access token is enough?